build-setup/vscode/using_vscode.md

```diff
@@ -0,0 +1,237 @@
+# Using Visual Studio Code with `ml-cpp`
+
+Previously, we explained how to [set up CLion for `ml-cpp`](../clion/using_clion.md). You may find, that CLion has some
+performance problems when indexing `ml-cpp` project: it may take very long to open the project or CLion may freeze and
+fail. In this case, [Visual Studio Code](https://code.visualstudio.com/) can be an alternative. When combined with
+`clangd`, it is significantly faster, while still being able to provide features like autocompletion, code navigation,
+static code analysis, etc.
+
+## Prerequisites
+
+In this tutorial, we assume that you have `gcc` or `clang` installed as described in the
+[CLion tutorial](../clion/using_clion.md). We also assume that you already created `compile_commands.json` file as
+described there since `clangd` also uses it.
+
+[Install `clangd`](https://clangd.llvm.org/installation.html). Version 11.0 had a significant performance improvement,
+so it is worse to go an extra mile and install the latest stable version. Install
+[Visual Studio Code](https://code.visualstudio.com/) with
+[clangd plugin](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd).
+
+In **Settings**, make sure *C_Cpp:Autocomplete* and *C_Cpp:Intelli Sense Engine* from the native C++ plugin of VS Code
+are disabled. Note that we don't want to disable the complete C/C++ Extension, since we are still going to use the
+debugging capabilities.
+
+Now for the `clangd` plugin, make sure that you have the correct path specified in **Settings** and *Clangd: Path*.
+*Clangd: Arguments* can look something like this:
+
+```text
+-log=verbose
+-pretty
+--header-insertion=iwyu
+--suggest-missing-includes
+-j=4
+--all-scopes-completion
+--background-index=0
+--clang-tidy
+--compile-commands-dir=./
+```
+
+## User tasks
+
+You can use [user tasks](https://code.visualstudio.com/docs/editor/tasks) to integrate VS Code with external tools. For
+example, re-format the project using a the docker container with correct clang-format version or build certain parts of
+the project in Debug mode. Later, we will see how we can use these tasks as dependencies for debugging.
+
+Here is an example of user tasks specified in `tasks.json`:
+
+```json
+{
+    "version": "2.0.0",
+    "tasks": [
+        {
+            "label": "Format code",
+            "type": "shell",
+            "command": "cd ${workspaceFolder} && dev-tools/docker/run_docker_clang_format.sh",
+            "problemMatcher": [],
+            "options": {
+                "env": {
+                    "PATH": "${env:PATH}",
+                    "CPP_SRC_HOME": "${env:CPP_SRC_HOME}"
```

Comment:
```suggestion
                    "CPP_SRC_HOME": "${workspaceFolder}"
```
If we have this option, I think it is better to use `${workspaceFolder}` here. For instance, if you have two copies of `ml-cpp` locally, this will make using VSCode with both of them easier.

Reply:
++ It makes more sense as `CPP_SRC_HOME` needn't be defined to build in VSCode